macroslib/src/python/mod.rs

```diff
@@ -0,0 +1,1219 @@
+        .clone())
+}
+
+fn method_name(method: &ForeignerMethod, src_id: SourceId) -> Result<syn::Ident> {
+    if method.variant == MethodVariant::Constructor {
+        parse("__new__", src_id)
+    } else {
+        let name = standard_method_name(method, src_id)?;
+        let name_str = name.to_string();
+        match name_str.as_ref() {
+            "to_string" => parse("__repr__", src_id),
+            _ => Ok(name),
+        }
+    }
+}
+
+fn self_type_conversion(
+    class: &ForeignerClassInfo,
+    method: &ForeignerMethod,
+    conv_map: &mut TypeMap,
+) -> Result<Option<TokenStream>> {
+    if let MethodVariant::Method(self_variant) = method.variant {
+        let self_type = &class
+            .self_desc
+            .as_ref()
+            .ok_or_else(|| {
+                DiagnosticError::new(
+                    class.src_id,
+                    class.span(),
+                    "Class have non-static methods, but no self_type",
+                )
+            })?
+            .self_type;
+        let self_type_ty = match self_variant {
+            SelfTypeVariant::Rptr => parse_type! {&#self_type},
+            SelfTypeVariant::RptrMut => parse_type! {&mut #self_type},
+            _ => parse_type! {#self_type},
+        };
+        Ok(Some(
+            generate_conversion_for_argument(
+                &conv_map.find_or_alloc_rust_type(&self_type_ty, class.src_id),
+                method.span(),
+                class.src_id,
+                conv_map,
+                "self",
+                true,
+            )?
+            .1,
+        ))
+    } else {
+        Ok(None)
+    }
+}
+
+fn has_any_methods(class: &ForeignerClassInfo) -> bool {
+    class.methods.iter().any(|m| {
+        if let MethodVariant::Method(_) = m.variant {
+            true
+        } else {
+            false
+        }
+    })
+}
+
+fn generate_conversion_for_argument(
+    rust_type: &RustType,
+    method_span: Span,
+    src_id: SourceId,
+    conv_map: &mut TypeMap,
+    arg_name: &str,
+    reference_allowed: bool,
+) -> Result<(Type, TokenStream)> {
+    let arg_name_ident: TokenStream = parse(arg_name, src_id)?;
+    if is_cpython_supported_type(rust_type) {
+        Ok((rust_type.ty.clone(), arg_name_ident))
+    } else if let Some((ty, conversion)) = if_exported_class_generate_argument_conversion(
+        rust_type,
+        conv_map,
+        &arg_name_ident,
+        method_span,
+        src_id,
+        reference_allowed,
+    )? {
+        Ok((ty, conversion))
+    } else if let Some(enum_info) = conv_map.is_this_exported_enum(&rust_type) {
+        let enum_py_mod: Ident = parse(&py_wrapper_mod_name(&enum_info.name.to_string()), src_id)?;
+        Ok((
+            parse_type!(u32),
+            quote! {
+                super::#enum_py_mod::from_u32(py, #arg_name_ident)?
+            },
+        ))
+    } else if let Some(inner) = ast::if_option_return_some_type(&rust_type) {
```

Comment:
The main idea is actually get rid of such kind of code,
and replace it with foreign_typemap! macro DSL: #173 ,
to improve extendability and reduce support cost.

But I am not sure that foreign_typemap mature enough to describe all conversation.
So let's leave it as it is.